From e70cc4b1f4fe33c20e679968ba32bdfda5520255 Mon Sep 17 00:00:00 2001
From: Moria Zuberi <moria.zuberi@tandemg.com>
Date: Tue, 06 Jul 2021 16:22:57 +0300
Subject: [PATCH] [SW-48525] gaudi2: erom: support pcie config

Add support for PCIe configuration in eROM code.
eROM code will be used in case there was a problem to config
the PCIe link with ROM code. I case of such issue, a dedicated
efuse will be used to sign that eROM code should be used.

Change-Id: Id1f47d6afd79853579138da857739980a2b109de
Signed-off-by: Moria Zuberi <moria.zuberi@tandemg.com>
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index 001e67b..7b9c38b 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -20,13 +20,29 @@
 
 EROM_APP_S := $(shell ls crt0_common_erom.S)
 EROM_APP_OBJECTS := $(EROM_APP_S:.S=.o)
-SOURCE_OBJECTS := erom_%.o
-SOURCE_OBJECTS += $(EROM_APP_OBJECTS)
+
+EROM_PCIE_DEPS := \
+	../../lib/common/pll.o \
+	../../lib/common/pci.o \
+	../../lib/common/uart.o \
+	../../app/ppboot/ppboot.o \
+	../../$(HL_DEV_ARCH)/$(HL_TARGET).o \
+	../../$(HL_DEV_ARCH)/../common/$(HL_TARGET).o \
+	../../$(HL_DEV_ARCH)/../common/common.o \
+	../../$(HL_DEV_ARCH)/common.o
 
 .PHONY: all
 all: erom_recovery erom_pcie erom_tpm
 
-erom_%: $(SOURCE_OBJECTS) $(EROM_LD_SCRIPT)
+erom_recovery: erom_recovery.o $(EROM_APP_OBJECTS) $(EROM_LD_SCRIPT)
+	@$(CC) $(LDFLAGS) -T $(EROM_LD_SCRIPT) -o $@  -Wl,--start-group $< $(LDLIBS) -Wl,--end-group
+	$(OBJCOPY) -O binary -R .note.gnu.build-id $@ $@.bin
+
+erom_pcie: erom_pcie.o $(EROM_APP_OBJECTS) $(EROM_PCIE_DEPS) $(EROM_LD_SCRIPT)
+	@$(CC) $(LDFLAGS) -T $(EROM_LD_SCRIPT) -o $@  -Wl,--start-group $< $(EROM_PCIE_DEPS) $(LDLIBS) -Wl,--end-group
+	$(OBJCOPY) -O binary -R .note.gnu.build-id $@ $@.bin
+
+erom_tpm: erom_tpm.o $(EROM_APP_OBJECTS) $(EROM_LD_SCRIPT)
 	@$(CC) $(LDFLAGS) -T $(EROM_LD_SCRIPT) -o $@  -Wl,--start-group $< $(LDLIBS) -Wl,--end-group
 	$(OBJCOPY) -O binary -R .note.gnu.build-id $@ $@.bin
 
diff --git a/app/ext_rom/erom_pcie.c b/app/ext_rom/erom_pcie.c
index 0f0818c..861a4f4 100644
--- a/app/ext_rom/erom_pcie.c
+++ b/app/ext_rom/erom_pcie.c
@@ -6,20 +6,19 @@
  */
 
 #include <stdint.h>
-#include <psoc_global_conf_regs.h>
-#include <hl_fw_if.h>
-#include <hl_boot_if.h>
-
-#define ADDR_SW_STATUS_2_REG	0xFCC60208
-
-void erom_pcie_config(void)
-{
-	/* Updating the SW status 2 register */
-	*(uint32_t *)(ADDR_SW_STATUS_2_REG) = 1;
-}
+#include <pci.h>
+#include <pll.h>
+#include <preboot.h>
 
 void erom_main_function(void)
 {
-	erom_pcie_config();
+	/*
+	 * Note: we assume CPU frequency is not changed while CPU is running,
+	 * so the udelay init is done once before we start.
+	 */
+	udelay_init();
+
+	hl_uart_init();
+	hl_pcie_init();
 }
 
